src/ansi-c/expr2c.cpp

```diff
@@ -165,7 +165,8 @@ std::string expr2ct::convert_rec(
   const c_qualifierst &qualifiers,
   const std::string &declarator)
 {
-  c_qualifierst new_qualifiers(qualifiers);
+  std::unique_ptr<c_qualifierst> clone = qualifiers.clone();
+  c_qualifierst &new_qualifiers = *clone;
```

Comment:
?

Reply:
Not entirely sure what you're asking, so I'll give a full answer, much of which you probably already know.
This code takes the existing qualifiers, adds some more to them and uses them on sub-elements. It doesn't update the qualifiers collection passed in. It therefore needs to make a copy of the qualifiers passed in. The qualifiers are passed by reference to a `c_qualifierst` but might be of a derived type, such as `java_qualifierst`. To allow this we need a virtual "copy" function, which is normally called `clone`. `clone` has to return by reference/pointer and so normally returns a `unique_ptr`. The unique pointer defines the scope of the variable to be this function and handles the destruction of the returned object but within this scope and in order to reduce code churn, it is possible to take a reference to the pointed to value and use it in this function.

Reply:
Thank you very much for the elaboration! Maybe I was aware of most of that, but having it written down helped nonetheless.

I don't think that having `expr2ct` call functions with a C-specific parameter type is ideal for re-use, but it should be possible to resolve most of that through double dispatch. Let me try to make detailed comments to that effect in the individual places.

Reply:
I agree that the type of the parameter to a virtual function intended to be overridden for various languages should be a non-language-specific qualifiers class and I can add such an abstract base class if you like, deriving `c_qualifierst` from it.

Reply:
My apologies that I have not actually come back with a double-dispatch solution. I have started working on one, noticed that this was way more work than I had initially thought, and then got dragged into another matter.

Here are my updated thoughts: I concur that there should be a non-language-specific parameter type. Since you can't do a virtual copy constructor, you might anyhow end up with the `clone` approach that you have already implemented. What I do hope to be avoidable is the `dynamic_cast`, which would be avoided by calling a virtual member of the parameter. All that is still a bit cumbersome, so you might actually consider implementation qualifiers via `irept`, which would make it easy to customise and would avoid all the clone/cast games.

Reply:
I like your thinking - double dispatch is always my preference over `dynamic_cast` - but I'm not sure the `dynamic_cast` can be avoided in this code - a good place to start looking is `java_qualifierst::operator==` - imagine how you'd implement that without `dynamic_cast`. Using `irept` doesn't avoid the problem, it just puts the class identifier into the `irep::id` field rather than RTTI - you'd still need to check it to do the comparison.

Reply:
Yes, all agreed. Overall it seems pretty painful to me and I am wondering whether you might just want a `type_qualifierst` that isn't exactly language-agnostic, but instead caters for several languages. For a start you might just put all the extra information that you added to `java_qualifierst` into `c_qualifierst` and reduce size of your patch a fair bit.

Reply:
I've added https://github.com/diffblue/cbmc/pull/1831/commits/2162f8ae8d546bc35cfcedc0f2e9434b7ac9d5bc, which brings in a base class, though it doesn't do everything we'd like. If you need more than this I'll need some more pointers.

Reply:
Our comments seem to have crossed. Would you prefer I swap all this out for a single type_qualifiert class now? I can see pros and cons both ways.

Reply:
I think (but you might convince me otherwise) that a single `type_qualifiert` class would make the code a lot simpler, yes.

Reply:
Oh no, it would definitely make the code simpler. But it would add Java specific stuff to a class used by CBMC. Pros and cons...
Does @peterschrammel have an opinion?